### DISCUSSION TITLE
How to delete nodes/edge onClick?

### DISCUSSION THREAD

**Discussion by MrRainesE** [xyflow/xyflow#2196] [Q&A]
I added a delete button and want to remove an node/edge  when this is clicked. i tried using nodes.splice(index, 1) but this does not update the state.
const onElementsRemove = useCallback(
    (elementsToRemove) => setElements((els) => removeElements(elementsToRemove, els)),
    []
  ); this works in older versions perfect, how can i do this in new vesion

**Comment by bcakmakoglu** [comment]
You can use `setNodes` to remove a node.

`const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes);`

```jsx
 <button
          onClick={() => {
            setNodes((n) => {
              n.shift();
              return [...n];
            });
          }}
        >
          Delete a node
</button>
```

[Here's a sandbox example.](https://codesandbox.io/s/sandpack-project-forked-ippx85?file=/App.js:913-1113)

There's different ways to do it as well, like using the `onNodesChange` callback like this:
```jsx
        <button
          onClick={() => {
            onNodesChange([{ type: "remove", id: nodes[0].id }]);
          }}
        >
          Delete a node
        </button>
```

**Reply by MrRainesE** [reply]
thanks, it works
